File: tisca/R/inference.R
# ----------------------------------------------------------------------------
# TISCA v2 -- inference.R
# Paired inference on the contrast D_j = L_{A,j} - L_{B,j}.
#
# C1 (plan Section 1.2 / spec Section 7.1): all inference on paired contrasts
# from common replications with common random numbers. NAs are dropped LISTWISE
# across the pair, with the count reported and a sensitivity note available.
# Per-column NA dropping is forbidden (v1 defect, plan Section 1.6).
#
# Routines provided:
#   contrast_from_columns(A, B)        -> list(D, dropped, n_kept)  listwise deletion
#   paired_t(D, alpha)                 -> one-sample t on D, estimate-first output
#   studentized_paired_bootstrap(D, B, alpha, mcse) -> bootstrap CI/rejection
#   paired_bootstrap_bca(D, B, alpha)  -> BCa CIs (auxiliary)
#   mcnemar_exact(x_A, x_B, alpha)     -> exact/binomial McNemar on paired 0/1 indicators
# ----------------------------------------------------------------------------

## ---- listwise deletion ------------------------------------------------------

#' Build the paired contrast with LISTWISE NA deletion across the pair.
#'
#' A replication j is kept for the contrast (A,B) only if BOTH L_{A,j} and L_{B,j}
#' are present and finite. The number dropped is reported. Per-column dropping is
#' not performed anywhere in this module.
#'
#' @param A numeric vector of replication-level losses for method A
#' @param B numeric vector of replication-level losses for method B
#' @return list(D = D_j, dropped = number dropped, n_kept = length(D),
#'              na_A = count of NAs in A, na_B = count of NAs in B,
#'              paired = TRUE)
contrast_from_columns <- function(A, B) {
  if (length(A) != length(B)) {
    stop("contrast_from_columns: A and B must be parallel (same replication index)")
  }
  ok   <- stats::complete.cases(A, B)
  D    <- A[ok] - B[ok]
  list(
    D       = D,
    dropped = sum(!ok),
    n_kept  = length(D),
    na_A    = sum(is.na(A) | is.nan(A)),
    na_B    = sum(is.na(B) | is.nan(B)),
    paired  = TRUE
  )
}

## ---- paired one-sample t ----------------------------------------------------

#' Paired one-sample t-test on D_j (df = J-1), the default fast inference.
#'
#' Returns estimate and Monte Carlo CI first and the p-value second, per IJDA #1e
#' (estimate-first reporting). Correctly removes the `paired=FALSE` defect of v1.
#'
#' @param D numeric vector of paired contrasts
#' @param alpha level (default 0.05)
#' @param mu null hypothesis value of E[D] (default 0)
#' @param alternative "two.sided","less","greater" (must match the planned mode)
#' @return list(estimate, mcse, ci_lower, ci_upper, t_stat, p_value, df, n)
paired_t <- function(D, alpha = 0.05, mu = 0, alternative = c("two.sided","less","greater")) {
  alternative <- match.arg(alternative)
  n <- length(D)
  if (n < 2) stop("paired_t: need at least 2 replications")
  dbar <- mean(D)
  s    <- stats::sd(D)
  mcse <- s / sqrt(n)
  tstat <- (dbar - mu) / mcse
  df <- n - 1
  p <- switch(alternative,
    two.sided = 2 * stats::pt(-abs(tstat), df),
    less      = stats::pt(tstat, df),
    greater   = stats::pt(tstat, df, lower.tail = FALSE)
  )
  crit <- stats::qt(1 - alpha / 2, df)
  half <- crit * mcse
  list(
    estimate   = dbar,
    mcse       = mcse,
    ci_lower   = dbar - half,
    ci_upper   = dbar + half,
    t_stat     = tstat,
    p_value    = p,
    df         = df,
    n          = n,
    method     = "paired t",
    dropped    = NA_integer_
  )
}

## ---- studentized paired bootstrap ------------------------------------------

#' Studentized paired bootstrap on D_j (recommended check for skewed D).
#'
#' The test statistic is T = (Dbar - mu)/se(D). The bootstrap re-estimates
#' Dbar_b and se_b for each resample and accumulates the studentized pivot
#' (Dbar_b - Dbar)/se_b. The reported CI and p-value are based on that pivot,
#' which is second-order accurate and handles the skew/long tails that break
#' the CLT-based t (P3-T4). Report the bootstrap MCSE of the p-value/CI bound.
#'
#' @param D numeric vector of paired contrasts
#' @param B bootstrap resamples (default 4999)
#' @param alpha level
#' @param mu null value
#' @param alternative sidedness
#' @param seed optional RNG seed for reproducibility
#' @param margin margin pass-through for the combined statistic (unused here, kept for API parity)
#' @return list(estimate, ci_lower, ci_upper, p_value, pivot_mcse, B, seed)
studentized_paired_bootstrap <- function(D, B = 4999L, alpha = 0.05, mu = 0,
                                         alternative = c("two.sided","less","greater"),
                                         seed = NULL, margin = NULL) {
  alternative <- match.arg(alternative)
  n <- length(D)
  if (n < 2) stop("studentized_paired_bootstrap: need >= 2 replications")
  if (!is.null(seed)) set.seed(seed)
  dbar <- mean(D)
  se   <- stats::sd(D) / sqrt(n)
  if (se <= 0) {
    # degenerate deterministic contrast
    p <- if (alternative == "two.sided") (dbar != mu) else (if (alternative == "less") dbar < mu else dbar > mu)
    return(list(estimate = dbar, ci_lower = dbar, ci_upper = dbar,
                p_value = as.numeric(p), pivot_mcse = 0, B = B, seed = seed,
                degenerate = TRUE))
  }
  # fast bootstrap resampling
  idx <- matrix(sample.int(n, size = n * B, replace = TRUE), nrow = B, ncol = n)
  db  <- rowMeans(matrix(D[idx], nrow = B, ncol = n))
  vb  <- apply(matrix(D[idx], nrow = B, ncol = n), 1,
               function(x) stats::sd(x) / sqrt(length(x)))
  piv <- (db - dbar) / vb
  piv <- piv[is.finite(piv)]
  B_eff <- length(piv)
  if (B_eff == 0) stop("studentized_paired_bootstrap: no finite bootstrap pivots")

  lv <- stats::quantile(piv, alpha / 2, type = 6)
  uv <- stats::quantile(piv, 1 - alpha / 2, type = 6)
  ci_lower <- dbar - uv * se
  ci_upper <- dbar - lv * se

  # p-value: fraction of pivots on the far side of the observed pivot (0 in pivot units)
  obs <- (mu - dbar) / se  # pivot centred on mu for the null
  pv <- switch(alternative,
    two.sided = mean(abs(piv) >= abs(obs)),
    less      = mean(piv <= obs),
    greater   = mean(piv >= obs)
  )
  # bootstrap MCSE of the p-value (rate estimate)
  p_mcse <- sqrt(pv * (1 - pv) / B_eff)
  list(estimate = dbar, ci_lower = ci_lower, ci_upper = ci_upper,
       p_value = pv, pivot_mcse = p_mcse, B = B, seed = seed,
       degenerate = FALSE, n_kept = n)
}

#' BCa bootstrap CIs on the mean (auxiliary; not the primary studentized path).
#' @param D numeric vector of paired contrasts
#' @param B resamples
#' @param alpha level
#' @param seed optional seed
#' @return list(estimate, ci_lower, ci_upper)
paired_bootstrap_bca <- function(D, B = 4999L, alpha = 0.05, seed = NULL) {
  if (!is.null(seed)) set.seed(seed)
  boot_mean <- function(data) {
    res <- replicate(B, mean(sample(data, replace = TRUE)))
    # BCa
    z0 <- stats::qnorm(mean(res < mean(data)))
    jack <- numeric(length(data))
    for (i in seq_along(data)) jack[i] <- mean(data[-i])
    a <- sum((mean(data) - jack)^3) / (6 * (sum((mean(data) - jack)^2))^1.5)
    z1 <- z0 + z0 + stats::qnorm(1 - alpha / 2)
    za <- z0 + (z0 + stats::qnorm(1 - alpha / 2)) / (1 - a * (z0 + stats::qnorm(1 - alpha / 2)))
    p_low <- stats::pnorm(z0 + (z0 - stats::qnorm(1 - alpha / 2)) / (1 - a * (z0 - stats::qnorm(1 - alpha / 2))))
    list(lo = stats::quantile(res, p_low, type = 6),
         hi = stats::quantile(res, 1 - p_low, type = 6),
         est = mean(data))
  }
  r <- boot_mean(D)
  list(estimate = r$est, ci_lower = r$lo[[1]], ci_upper = r$hi[[1]])
}

## ---- McNemar for paired binary rows (4, 12) ----------------------------------

#' Exact/binomial McNemar test on paired binary indicators.
#'
#' For two binary columns x_A, x_B (e.g. ATE coverage indicators), only the
#' discordant pairs (1,0) and (0,1) carry information. The test is the sign test
#' on the discordant pairs: b = count of (A=1,B=0), c = count of (A=0,B=1),
#' H0: b/(b+c) = 0.5, with an exact binomial two-sided p-value (R's binom.test).
#' A mid-p correction is optionally applied (small-sample exact calibration).
#'
#' @param x_A 0/1 replicate-level indicator for A
#' @param x_B 0/1 replicate-level indicator for B
#' @param alpha level
#' @param alternative sidedness of the difference in coverage
#' @param midp apply the mid-p correction
#' @return list(b, c, n_discordant, p_value, estimate, ci_lower, ci_upper, method)
mcnemar_exact <- function(x_A, x_B, alpha = 0.05,
                          alternative = c("two.sided","less","greater"),
                          midp = FALSE) {
  alternative <- match.arg(alternative)
  ok  <- stats::complete.cases(x_A, x_B)
  x_A <- x_A[ok]; x_B <- x_B[ok]
  b <- sum(x_A == 1 & x_B == 0)
  c <- sum(x_A == 0 & x_B == 1)
  n_disc <- b + c
  est <- if (n_disc > 0) b / n_disc else NA_real_
  if (n_disc == 0) {
    return(list(b = b, c = c, n_discordant = 0, p_value = 1,
                estimate = NA_real_, ci_lower = NA_real_, ci_upper = NA_real_,
                method = "McNemar exact (no discordant pairs)"))
  }
  # p-value under H0: X ~ Bin(n_disc, 0.5). Two-sided p via doubling the smaller tail.
  # (Also handles mid-p correction by subtracting half the point mass at b.)
  pv <- switch(alternative,
    two.sided = 2 * pmin(stats::pbinom(b, n_disc, 0.5),
                         stats::pbinom(b - 1, n_disc, 0.5, lower.tail = FALSE)),
    less      = stats::pbinom(b, n_disc, 0.5),
    greater   = stats::pbinom(b - 1, n_disc, 0.5, lower.tail = FALSE)
  )
  if (alternative == "two.sided") pv <- pmin(pv, 1)  # clamp the doubled tail
  if (midp) {
    pmf <- stats::dbinom(b, n_disc, 0.5)
    pv  <- pmin(pmax(pv - 0.5 * pmf, 0), 1)
  }
  ci <- stats::binom.test(b, n_disc, 0.5, conf.level = 1 - alpha)$conf.int
  list(b = b, c = c, n_discordant = n_disc, p_value = pv,
       estimate = est, ci_lower = ci[[1]], ci_upper = ci[[2]],
       dropped = sum(!ok), method = "McNemar exact (sign test on discordant pairs)")
}

File: tisca/python/tisca/inference.py
"""Paired inference — `inference.py`.

Implements §7.1 of `docs/tisca_v2_spec.md`:
* **paired one-sample t** on ``D_j`` (df ``J-1``) as the fast default;
* **studentized paired bootstrap** (B resamples, its MCSE reported) as the
  recommended check for non-trivial skewness;
* **McNemar exact / binomial** test for paired binary indicators (rows 4, 12).

All contrasts are formed on common replications with **listwise deletion across
the pair** (convention 3); per-column NA dropping is forbidden.
"""

from __future__ import annotations

import numpy as np
from scipy.stats import t as _t, binom as _binom

from . import validate, estimands

__all__ = [
    "paired_t",
    "paired_t_test",
    "studentized_paired_bootstrap",
    "paired_bootstrap_ci",
    "mcnemar_exact",
    "mcnemar_with_continuity",
    "contrast_summary",
]

_DEFAULT_B = 4999


def paired_t(D, mu0: float = 0.0, *, alternative: str = "two-sided") -> dict:
    """One-sample t on the paired contrasts ``D_j``.

    Returns estimate, SE, t statistic, degrees of freedom, and a ``p_value``.
    ``alternative`` is 'two-sided', 'less' (``H1: theta < 0``, lower-is-better) or
    'greater'.
    """
    D = np.asarray(D, dtype=float)
    J = D.size
    if J < 2:
        raise validate.ValidationError(f"paired_t needs >= 2 replications, got {J}.")
    mean = float(np.mean(D))
    sd = float(np.std(D, ddof=1))
    se = sd / np.sqrt(J)
    df = J - 1
    if sd == 0.0:
        # Degenerate contrasts have no sampling uncertainty on the difference.
        if alternative == "two-sided":
            p = 0.0 if abs(mean - mu0) > 0 else 1.0
        elif alternative == "less":
            p = 0.0 if mean < mu0 else 1.0
        elif alternative == "greater":
            p = 0.0 if mean > mu0 else 1.0
        else:
            raise validate.ValidationError(f"Unknown alternative {alternative!r}.")
        tstat = 0.0
    else:
        tstat = (mean - mu0) / se
        if alternative == "two-sided":
            p = 2.0 * _t.sf(abs(tstat), df=df)
        elif alternative == "less":
            p = _t.cdf(tstat, df=df)
        elif alternative == "greater":
            p = _t.sf(tstat, df=df)
        else:
            raise validate.ValidationError(f"Unknown alternative {alternative!r}.")
    return {
        "estimate": mean,
        "sd": sd,
        "se": se,
        "mcse": se,
        "t": tstat,
        "df": df,
        "J": J,
        "p_value": float(p),
        "alternative": alternative,
    }


paired_t_test = paired_t


def studentized_paired_bootstrap(
    D,
    *,
    mu0: float = 0.0,
    B: int = _DEFAULT_B,
    alternative: str = "two-sided",
    seed: int | None = None,
    bootstrap_indices: np.ndarray | None = None,
) -> dict:
    """Studentized paired bootstrap over the mean ``D_bar``.

    For each resample the studentized statistic is
    ``T_b = (D_bar_b - D_bar) / se_b``; the confidence interval is built from the
    quantiles of ``T_b`` and the p-value from their centred position. Returns an
    ``estimate``, a ``(1-level)`` ``ci``, a ``p_value``, and the bootstrap MCSE.
    """
    D = np.asarray(D, dtype=float)
    J = D.size
    if J < 2:
        raise validate.ValidationError("studentized bootstrap needs >= 2 replications.")
    if alternative not in ("two-sided", "less", "greater"):
        raise validate.ValidationError(f"Unknown alternative {alternative!r}.")
    mean = float(np.mean(D))
    sd = float(np.std(D, ddof=1))
    if sd == 0.0:
        # Degenerate: all bootstrap studentized statistics are undefined; fall back
        # to the deterministic point.
        base = dict(
            estimate=mean, sd=sd, se=0.0, mcse=0.0, p_value=None, ci=(mean, mean),
            B=B, seed=seed, degenerate=True,
        )
        return base

    se_obs = sd / np.sqrt(J)

    if bootstrap_indices is not None:
        idx = np.asarray(bootstrap_indices, dtype=int)
        if idx.ndim != 2 or idx.shape[1] != B or idx.shape[0] != J:
            raise validate.ValidationError(
                f"bootstrap_indices must have shape (J={J}, B={B}); got {idx.shape}."
            )
    else:
        rng = np.random.default_rng(seed)
        idx = rng.integers(0, J, size=(J, B))

    shape = idx.shape
    D_boot = D[idx]  # (J, B)
    D_bar_b = D_boot.mean(axis=0)
    sd_b = D_boot.std(axis=0, ddof=1)
    with np.errstate(divide="ignore", invalid="ignore"):
        se_b = sd_b / np.sqrt(J)
        T_b = (D_bar_b - mean) / se_b
    T_b = np.nan_to_num(T_b, nan=0.0, posinf=10.0, neginf=-10.0)

    level = 0.95
    if alternative == "two-sided":
        lo_q = np.quantile(T_b, 1 - (1 - level) / 2)
        hi_q = np.quantile(T_b, (1 - level) / 2)
        lower = mean - lo_q * se_obs
        upper = mean - hi_q * se_obs
        t_obs = (mean - mu0) / se_obs
        p = 2.0 * min(float(np.mean(T_b <= -abs(t_obs))), float(np.mean(T_b >= abs(t_obs))), 1.0)
    elif alternative == "less":
        hi_q = np.quantile(T_b, 1 - level)
        upper = mean - hi_q * se_obs
        lower = -np.inf
        t_obs = (mean - mu0) / se_obs
        p = float(np.mean(T_b <= t_obs))
    else:  # greater
        # One-sided LOWER bound [mean - q_{1-a}(T) se, inf). The upper quantile
        # is the right one here: using q_{a}(T) (which is negative) put the
        # "lower" bound ABOVE the point estimate.
        lo_q = np.quantile(T_b, level)
        lower = mean - lo_q * se_obs
        upper = np.inf
        t_obs = (mean - mu0) / se_obs
        p = float(np.mean(T_b >= t_obs))
    p = min(1.0, max(0.0, p))

    return dict(
        estimate=mean,
        sd=sd,
        se=se_obs,
        mcse=se_obs,
        p_value=float(p),
        ci=(float(lower), float(upper)),
        B=B,
        seed=seed,
        degenerate=False,
    )


paired_bootstrap_ci = studentized_paired_bootstrap


def _discordant_counts(a, b):
    """McNemar table cells: n01, n10."""
    a = np.asarray(a, dtype=int)
    b = np.asarray(b, dtype=int)
    n01 = int(np.sum((a == 0) & (b == 1)))
    n10 = int(np.sum((a == 1) & (b == 0)))
    return n01, n10


def mcnemar_exact(a, b) -> dict:
    """Exact (binomial) McNemar test on paired binary indicators.

    Under ``H0`` the discordant pairs are equally likely; conditioning on
    ``m = n01 + n10``, ``n10 ~ Binomial(m, 1/2)``. Two-sided p-value uses the
    doubled one-sided tail capped at 1.
    """
    a = np.asarray(a)
    b = np.asarray(b)
    if a.size != b.size:
        raise validate.ValidationError("McNemar requires equal-length paired indicators.")
    validate.validate_binary(a, name="a")
    validate.validate_binary(b, name="b")
    n01, n10 = _discordant_counts(a, b)
    m = n01 + n10
    if m == 0:
        p = 1.0
    else:
        k = min(n01, n10)
        p_one = _binom.cdf(k, n=m, p=0.5)
        p = min(1.0, 2.0 * p_one)
    return {
        "n": a.size,
        "n01": n01,
        "n10": n10,
        "m_discordant": m,
        "p_value": float(p),
        "exact": True,
    }


def mcnemar_with_continuity(a, b) -> dict:
    """McNemar with Yates continuity correction (chi-square approximation)."""
    from scipy.stats import chi2 as _chi2

    a = np.asarray(a)
    b = np.asarray(b)
    validate.validate_binary(a, name="a")
    validate.validate_binary(b, name="b")
    n01, n10 = _discordant_counts(a, b)
    diff = abs(n10 - n01)
    m = n01 + n10
    if m == 0:
        chi2 = 0.0
    else:
        chi2 = ((diff - 1.0) ** 2) / float(m)
    p = float(_chi2.sf(chi2, df=1))
    return {"n01": n01, "n10": n10, "m_discordant": m, "chi2": chi2, "p_value": p}


def contrast_summary(A, B, *, name: str = "contrast", paired: bool = True) -> dict:
    """Build a full per-contrast summary on the common-replication contrast ``D``.

    Applies listwise deletion across the pair (convention 3), then reports the
    paired estimate, MCSE, CI, and the number of replications dropped due to
    missing values.
    """
    a, b, n_dropped = validate.validate_contrast_pair(A, B, name=name)
    D = a - b
    res = paired_t(D)
    res["n_dropped"] = n_dropped
    res["D"] = D
    ci = estimands.nc_ci(res["estimate"], res["se"], level=0.95, df=res["df"])
    res["ci"] = ci
    res["name"] = name
    res["paired"] = paired
    return res
